File: Beecrowd_Parte3/Jupiter_Ataca_1356.c
#include <stdio.h>

#define MAXN 100010

typedef long long ll;

ll seg[4*MAXN], lazy[4*MAXN], pot[MAXN], inv[MAXN], vet[MAXN], B, P;

ll fexp(ll base, ll exp) {
    ll ans = 1;
    while(exp) {
        if(exp & 1) ans = (ans*base)%P;
        base = (base*base)%P;
        exp >>= 1;
    }
    return ans;
}

void build(int node, int start, int end) {
    lazy[node] = -1;
    if(start == end) {
        seg[node] = vet[start];
    } else {
        int mid = (start+end)>>1;
        build(2*node, start, mid);
        build(2*node+1, mid+1, end);
        seg[node] = (seg[2*node]*pot[end-mid] + seg[2*node+1])%P;
    }
}

void update(int node, int start, int end, int l, int r, ll val) {
    if(lazy[node] != -1) {
        seg[node] = (lazy[node]*pot[end-start+1])%P;
        if(start != end) {
            lazy[2*node] = lazy[node];
            lazy[2*node+1] = lazy[node];
        }
        lazy[node] = -1;
    }
    if(start > r || end < l) return;
    if(start >= l && end <= r) {
        seg[node] = (val*pot[end-start+1])%P;
        if(start != end) {
            lazy[2*node] = val;
            lazy[2*node+1] = val;
        }
        return;
    }
    int mid = (start+end)>>1;
    update(2*node, start, mid, l, r, val);
    update(2*node+1, mid+1, end, l, r, val);
    seg[node] = (seg[2*node]*pot[end-mid] + seg[2*node+1])%P;
}

ll query(int node, int start, int end, int l, int r) {
    if(start > r || end < l) return 0;
    if(lazy[node] != -1) {
        seg[node] = (lazy[node]*pot[end-start+1])%P;
        if(start != end) {
            lazy[2*node] = lazy[node];
            lazy[2*node+1] = lazy[node];
        }
        lazy[node] = -1;
    }
    if(start >= l && end <= r) return seg[node];
    int mid = (start+end)>>1;
    ll p1 = query(2*node, start, mid, l, r);
    ll p2 = query(2*node+1, mid+1, end, l, r);
    return (p1*pot[end-mid] + p2)%P;
}

int main() {
    int L, N;
    while(scanf("%lld %lld %d %d", &B, &P, &L, &N), B || P || L || N) {
        for(int i = 0; i <= L; i++) vet[i] = 0;
        pot[0] = 1;
        for(int i = 1; i <= L; i++) pot[i] = (pot[i-1]*B)%P;
        build(1, 0, L-1);
        for(int i = 0; i < N; i++) {
            char op;
            int a, b;
            scanf(" %c %d %d", &op, &a, &b);
            if(op == 'E') {
                update(1, 0, L-1, a-1, a-1, b);
                vet[a-1] = b;
            } else {
                printf("%lld\n", query(1, 0, L-1, a-1, b-1));
            }
        }
        printf("-\n");
    }
    return 0;
}

File: Beecrowd_Parte3/Jupiter_Ataca_1356.java
import java.util.Scanner;

public class Jupiter_Ataca_1356 {
    static class SegmentTree {
        long[] t, a;
        int p, b;

        SegmentTree(int n, int b, int p) {
            t = new long[n << 2];
            a = new long[n + 1];
            this.p = p;
            this.b = b;
        }

        void build(int node, int start, int end) {
            if (start == end) {
                t[node] = a[start];
            } else {
                int mid = (start + end) >> 1;
                build(node << 1, start, mid);
                build(node << 1 | 1, mid + 1, end);
                t[node] = (t[node << 1] * b + t[node << 1 | 1]) % p;
            }
        }

        void update(int node, int start, int end, int idx, long val) {
            if (start == end) {
                a[idx] = val;
                t[node] = val;
            } else {
                int mid = (start + end) >> 1;
                if (start <= idx && idx <= mid) {
                    update(node << 1, start, mid, idx, val);
                } else {
                    update(node << 1 | 1, mid + 1, end, idx, val);
                }
                t[node] = (t[node << 1] * b + t[node << 1 | 1]) % p;
            }
        }

        long query(int node, int start, int end, int l, int r) {
            if (r < start || end < l) {
                return 0;
            }
            if (l <= start && end <= r) {
                return t[node] * power(b, r - end) % p;
            }
            int mid = (start + end) >> 1;
            long p1 = query(node << 1, start, mid, l, r);
            long p2 = query(node << 1 | 1, mid + 1, end, l, r);
            return (p1 + p2) % p;
        }

        long power(long a, long b) {
            long res = 1;
            while (b > 0) {
                if ((b & 1) == 1) {
                    res = (res * a) % p;
                }
                b >>= 1;
                a = (a * a) % p;
            }
            return res;
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        while (true) {
            int b = sc.nextInt();
            int p = sc.nextInt();
            int l = sc.nextInt();
            int n = sc.nextInt();
            if (b == 0 && p == 0 && l == 0 && n == 0) {
                break;
            }
            SegmentTree st = new SegmentTree(l, b, p);
            for (int i = 0; i < n; i++) {
                char type = sc.next().charAt(0);
                int x = sc.nextInt();
                int y = sc.nextInt();
                if (type == 'E') {
                    st.update(1, 1, l, x, y);
                } else {
                    System.out.println(st.query(1, 1, l, x, y));
                }
            }
            System.out.println("-");
        }
        sc.close();
    }
}